### ISSUE TITLE
Navigation - React Navigation 'focus' listener is not triggered the first time

### ISSUE THREAD

**Issue by alessioemireni** [kirillzyusko/react-native-bundle-splitter#25]
**Describe the bug**
If I wrap the screen with `register()` method the React Navigation 5 'focus' listener is not triggered the first time. This happens also with React Component or PureComponent not just with Functional Component.

**Code snippet**
```
...
useEffect(() => {
    const unsubscribe = navigation.addListener('focus', () => {
      // The screen is focused
      // Call any action
      console.log("FOCUS DETAILS"); // Not triggered the first time
    });
    // Return the function to unsubscribe from the event so it gets removed on unmount
    return unsubscribe;
  }, [navigation]);
...
```

**Repo for reproducing**
Here you can find an example repository --> https://github.com/alessioemireni/react-native-bundle-splitter-example. It's a fork of your example repository.

**To Reproduce**
1. Open the App
2. See the console log. The 'Focus Home' is printed.
3. Click on 'Go to details'
4. See the console log. The 'Focus Details' log is not printed.
5. Go Back.
6. See the console log. The 'Focus Home' is printed.
7. Click on 'Go to details'
8. See the console log. The 'Focus Details' log is now printed.

**Expected behavior**
The React Navigation 5 'focus' listener should be triggered.

**Smartphone (please complete the following information):**
 - Desktop OS: [MacOS 10.15.7]
 - Device: [iPhone XR, Google Pixel 2 XL]
 - OS: [iOS 14.2, Android 11]
 - JS enginge [JSC, Hermes]
 - Library version [2.0.0]
 - React Native: [0.62.2]
 - React Navigation: [5.9.2]

**Comment by kirillzyusko** [comment]
Hi @alessioemireni 
Thanks for the detailed issue!
I will have a look on that.

**Comment by kirillzyusko** [comment]
Hi @alessioemireni 
I did some research around a problem and here is my results of the investigation.

First of all I think it's expected behavior. Let me explain why.
The basic idea behind this library is deferred components loading. And this loading is made in async style. What happens, when you wrap your screen within a `register` function? This function is simple HOC component, that represent stub of a screen during async loading. I. e. when you trigger a navigation to your screen, navigation will be triggered to a stub screen (which will later display real one once loading is finished).

Now let's go step by step of what happens under hood.

1. You trigger a navigation to Details screen.
2. Stub screen is displayed (and in this time it tries to load real one).
3. New screen is displayed (stub) and `react-navigation` triggers/emit `focus` event (but at this time you haven't loaded real screen yet -> you don't have any listeners in Details screen, because it actually wasn't loaded).
4. New screen is displayed (real one) but you don't receive `focus` event, since it was emitted before (when you didn't setup listeners).

I think at that time a new question will emerge in your head: "Why it works second time?". The answer is quite simple: once screen is loaded the library will automatically put it into cache and in the next time, when this screen will be used it will use version from cache, and an access to cache can be made synchronously (i. e. screen will be mounted in sync manner).

So answering on your question - If you want to receive `focus` event in Details for the first time too, you will need to use `preload` API. Look at the diff below:

```diff
diff --git a/src/screens/Details/index.tsx b/src/screens/Details/index.tsx
index 93c0670..ff82b19 100644
--- a/src/screens/Details/index.tsx
+++ b/src/screens/Details/index.tsx
@@ -1,10 +1,11 @@
 // import React from 'react';
-// import {register} from 'react-native-bundle-splitter';
+import {register} from 'react-native-bundle-splitter';
 
-import View from './View';
+// import View from './View';
 
 export default register({
+  name: 'Details',
   loader: () => import('./View'),
 });
 
-export default View;
+// export default View;
diff --git a/src/screens/Home/View.tsx b/src/screens/Home/View.tsx
index eaf3743..4ddbc53 100644
--- a/src/screens/Home/View.tsx
+++ b/src/screens/Home/View.tsx
@@ -1,6 +1,7 @@
 import React, { useCallback, useEffect } from 'react';
 import {Button, View, Text} from 'react-native';
 import { useFocusEffect } from '@react-navigation/native';
+import { preload } from 'react-native-bundle-splitter';
 
 // @ts-ignore
 function HomeScreen({navigation}) {
@@ -22,6 +23,10 @@ function HomeScreen({navigation}) {
     return unsubscribe;
   }, [navigation]);
 
+  useEffect(() => {
+    preload().component('Details');
+  });
+
   return (
     <View style={{flex: 1, alignItems: 'center', justifyContent: 'center'}}>
       <Text>Home Screen</Text>
```

I simply added a name for a screen (and commented out unnecessary code) and inside of `Home` screen I added preloading of the component. I put it in `useEffect`, but you can make it as you wish.

I highly recommend to use `preload` API. It eliminates the difference between synchronous and asynchronous code and helps make library integration almost seamless. The most common use cases for using the preloading API are: 
- loading additional screens during login (protected screens, that can be viewed only by authenticated user).
- loading some screens during application start (useful when you want to reduce the waiting time on the native splash screen and when you have a screen with a spinner after the native splash screen (maybe you are downloading some additional data or something like that)).

But I feel like I'm getting distracted from the topic 🙂 
If you use the preload API, there will be no problems.

**Comment by alessioemireni** [comment]
@kirillzyusko thanks for your detailed explanation. I will check your patch and replace the register method with preload in my main project. 

Why the behaviour is different with React Navigation 4? The focus was triggered the first time.

Another question. Why does the `useFocusEffect` react navigation hook work instead? The code snippet is commented in the repository example.
```
...
useFocusEffect(
     useCallback(() => {
       console.log("FOCUS HOOK DETAILS") --> triggered the first time
     }, [])
  );
...
```

**Comment by kirillzyusko** [comment]
@alessioemireni it's hard to answer. A lot of things were changed in 5.x comparison to 4.x...

Also it's hard to answer, why one thing works and other not. Need to debug and dive into react-navigation implementation in order to answer on this question 🙂

**Comment by alessioemireni** [comment]
We decided to use an Hoc with useFocusEffect react navigation Hook due to the fact that that all Screens in main project use focus Listener. I'll close the issue.
